### ISSUE TITLE
Regression: mypy thinks line is unreachable in 0.780 upwards, while (correctly) not showing a warning in 0.770 and below

### ISSUE THREAD

**Issue by rggjan** [python/mypy#9457]
This code:

```
class Foo:
    def __init__(self) -> None:
        self.dirty = False


def test_source() -> None:
    foo = Foo()

    def bar() -> None:
        foo.dirty = True

    assert foo.dirty is False

    bar()

    assert foo.dirty is True
    print("unreachable?")
```

[produces the warning](https://mypy-play.net/?mypy=0.780&python=3.8&flags=warn-unreachable&gist=1ad7dad65b67adeea86840225e823ae5): `main.py:17: error: Statement is unreachable`, when running `mypy --warn-unreachable test.py` on *0.780*.

Running the same [through mypy 0.770](https://mypy-play.net/?mypy=0.770&python=3.8&flags=warn-unreachable&gist=1ad7dad65b67adeea86840225e823ae5) works fine and doesn't produce a warning.

(Also, running the example with `pytest -s test.py` prints `unreachable?`, because obviously this line is not unreachable when running the code in `python`).

**Comment by gvanrossum** [comment]
Reproduced. This is very unfortunate. I imagine using git bisect one could find out what PR caused this. Are you interested in helping out? It would mean a lot.

**Comment by henribru** [comment]
I have bisected this to https://github.com/python/mypy/commit/65e3a89890518aa2ce7bf2e37579c927267c3851

**Comment by gvanrossum** [comment]
Thanks! I note that the bug already existed for enum literals though:
```
from enum import Enum

class Color(Enum):
    RED = 'red'
    BLUE = 'blue'

class Foo:
    def __init__(self) -> None:
        self.dirty = Color.RED


def test_source() -> None:
    foo = Foo()

    def bar() -> None:
        foo.dirty = Color.BLUE

    assert foo.dirty is Color.RED

    bar()

    assert foo.dirty is Color.BLUE
    print("unreachable?")
```
So I think the root cause is deeper. Presumably the code touched by the commit you dug up has some relevance. Do you feel like digging in deeper?

**Comment by abrahamfrancis** [comment]
I bisected the enum issue to e818a96917a5111c4a5b901d49aa6b36de9700d3

**Comment by gvanrossum** [comment]
Thanks! It seems the root problem here is that the binder (i.e., the part of mypy that tracks how control flow affects the assumed types of variables) can track instance variable types, and its heuristic to decide when it can no longer reliably know a variable's type is flawed. There are examples of this that do not involve enums or bool:
```
from typing import Union

class C:
    v: Union[int, str]
    def flop(self) -> None:
        self.v = "abc"

def f() -> None:
    c = C()
    c.v = 42
    reveal_type(c.v)  # int, correctly
    c.flop()
    reveal_type(c.v)  # int, wrongly!
```
Presumably we can construct similar examples using a local variable and a nested function that modifies the variable using `nonlocal`. (If someone can post one, that'd be great.)

This looks pretty difficult to fix without throwing the baby out with the bath water -- surely most function calls do *not* have side effects like this.

FWIW I don't expect the second reveal_type() to print `str`; but it would be nice if it printed `Union[int, str]`.

**Comment by nicoddemus** [comment]
To add another example, just noticed this while writing some tests, with mypy stopping pointing out problems because it believes the statements below an `assert` are unreachable:

```python
from typing import Optional

class C:
    r: Optional[int] = None

    def compute(self, x: int) -> None:
        self.r = x * 10


def test_c() -> None:
    c = C()

    assert c.r is None

    c.compute(10)
    assert c.r is not None
    assert c.r == 10
	c.r = "foo"  # mypy doesn't detect this.
```

I get the following in both `0.780` and `0.800`:

```
λ mypy foo.py
foo.py:16: error: Statement is unreachable  [unreachable]
Found 1 error in 1 file (checked 1 source file)
```

(The unreachable statement in the message above is `assert c.r is not None`).

This is similar to @gvanrossum's example above, but showcases the problem in unit-testing functions that change instance properties.

**Comment by hauntsaninja** [comment]
As a workaround, you could reset what the binder thinks it knows using something like `if TYPE_CHECKING: c = C()`

Edit: you can actually just do a dummy assignment `c = c`

**Comment by nicoddemus** [comment]
Thanks for the tip @hauntsaninja. 👍

**Comment by terencehonles** [comment]
@hauntsaninja do you know if there's a plan for fixing this or at least a discussion happening about this. I'm not sure if this would affect other type checkers and if there needs to be a consensus or at least a standardization of behavior.

One of my coworkers ran into this previously and we just disabled typing in the test he was writing and how I came to submit #10291 since he wasn't *trying* to type the test in question (we run with unreachable warnings on) so we had to disable the file until #10291 is released.

I definitely understand @gvanrossum's point that in general many functions don't modify any/enough state to run into this and it would create a lot of warnings if binder lost tracking this state.

Part of the reason I'm asking is because the suggestion you provide works ok in a trivial case like this, but I'm assuming you'd actually need to fill out the whole constructor if there are required arguments and the whole `if TYPE_CHECKING` block is a bit wordy. I'm wondering if it might make sense to have a function as part of `typing_extensions` that serves as a marker for this functionality but doesn't require "reinitializing" a variable explicitly.

Another option might be to also allow type checking to happen after what mypy thinks is unreachable, but I'm assuming that's actually unrealistic because mypy would need to somehow backtrack on what it thinks the types are so that it can even offer suggestions past the point it's saying it needs to bail.

**Comment by erictraut** [comment]
You asked about other type checkers. Pyright also makes the assumption that the type of a member access expression (e.g. `foo.dirty`) is not modified as a side effect by a function call. This is generally a good assumption, and it allows for better type inference and type narrowing in the vast majority of cases. This same assumption is used by type checkers in other languages like TypeScript. 

Unlike mypy, pyright does not report errors for unreachable code, nor does it stop analyzing code that is potentially unreachable based on type narrowing evaluation. For those reasons, the code examples above do not generate any errors in pyright.

In my experience, the pattern shown in the samples above (where a non-local variable is accessed, then modified as a side effect of a function call, then accessed again) comes up mainly in unit tests but rarely in non-test code.

You mentioned that the workaround provided by @hauntsaninja can get wordy for non-trivial constructors. Here's a variant that is simpler and works with both mypy and pyright to reset the narrowed type for the expression. You can use it with or without an `if TYPE_CHECKING:` check.

```python
    foo = foo
```

**Comment by hauntsaninja** [comment]
Another option is to turn off `--warn-unreachable` for tests ([similar example](https://github.com/hauntsaninja/boostedblob/blob/46c4f99605f77dc55c5132808b4b1b57e93d9c51/mypy.ini#L6)).

**Comment by terencehonles** [comment]
> Another option is to turn off `--warn-unreachable` for tests ([example of different mypy settings for tests](https://github.com/hauntsaninja/boostedblob/blob/46c4f99605f77dc55c5132808b4b1b57e93d9c51/mypy.ini#L6)).

Unfortunately @hauntsaninja are tests are not separated like this and scattered through the source tree. I can see if we'd be OK moving them, but I was aware I could change the settings. I just didn't have a good way to do it since mypy config needs to have full module matches and can't just be a regex (which is how our pytest's test discovery finds all the tests).

Thanks for the input @eric-wieser , I had realized after my comment you could probably define:

```python
T = TypeVar('T')
def reset(value: T) -> T: return value
```

but the simple assignment still beats it for simplicity. I agree that I've never encountered a time that I needed this for a non test code since I generally have fixed what mypy understood of the types. I guess the only place I've run into something adjacent to this is when re-using a variable and changing its type. That would be solved by allowing binder to understand that it should treat the variable as a new variable (and for the code I've started to add typing to I've just added the new variable name when needed).

**Comment by ShadowLNC** [comment]
Potentially related: #14010

**Comment by Derekmod** [comment]
The pyre approach is to never type narrow object attributes, and that works pretty well. You just have to use a local variable if you want narrowing. E.g.
```python
if (value := self.value) is not None:
    # value is nonnull here
```

I would really appreciate if this could be added at least as a flag

**Comment by ddanier** [comment]
One could argue that attribute narrowing is not really practical at all. See this example:

```python
import random
from typing import reveal_type

class Foo:
    @property
    def not_narrowable(self) -> bool:
        return random.choice((True, False))


x = Foo()

assert x.not_narrowable is True

reveal_type(x.not_narrowable)
```
(I know this code will fail at runtime about 50% of the time, but let's not argue about this code being useful. You could easily create a more realistic example - this here is just a very minimalistic one.)

All type checkers I tried will return `Literal[True]` as the revealed type (checked on `mypy`, `ty`, `pyright`, `pyrefly`). I think this short example shows how broken narrowing the type on object attributes behaviour **might** be.

Using a `property` might be the extreme (and obvious) case, but there is absolutely no guarantee for normal attribute values to keep their type after the assertion as well (think about calling methods with such side effects like mentioned previously or async code/threads sharing access to the same object for example).
